### PR TITLE
Configuration Handler for pyroSAR

### PR DESCRIPTION
**Pull request by ibaris**
ConfigHandler is a configuration handler for pyroSAR. It is intended to be called by a class's "__init__" and set or get the configuration parameters throughout an entire package.
The primary goal with ConfigHandler is to load a single, consistent configuration environment to be passed amongst ALL objects within a package.

ConfigHandler is a SINGLETON, meaning once instantiated, THE SAME OBJECT will be returned to every class object calling it.

Examples
```
>>> from pyroSAR import ConfigHandler
>>> config = ConfigHandler()
>>> config.add_section('SNAP')
>>> config.set('SNAP', 'etc', 'C:\Users\user\dir_1')
>>> config.set('SNAP', 'auxdata', 'C:\Users\user\dir_2')
>>> print config
        "Class    : Config
         Path     : C:\Users\ibari\.pyrosar\ConfigHandler.ini
         Sections : 1

         Contents:
           Section: SNAP
	            x etc :: C:\Users\user\dir_1 :: <type 'str'>
	            x auxdata :: C:\Users\user\dir_2 :: <type 'str'>"

>>> config.SNAP
        {'auxdata': 'C:\\Users\\user\\dir_2', 'etc': 'C:\\Users\\user\\dir_1'}

>>> config.SNAP['etc']
        'C:\\Users\\user\\dir_1'

>>> config.SNAP['auxdata']
        'C:\\Users\\user\\dir_2'
		
>>> config.sections
	['SNAP']
		
>>> config.keys('SNAP')
	['etc', 'auxdata', 'properties', 'auxdatapath']
	
>>> config.open()
	'The config.ini file will be opened'
```

Due to an bug of (only what I guess) configparser in python 3 the build failes:
```
==================================== ERRORS ====================================
_________________ ERROR collecting pyroSAR/tests/test_snap.py __________________
pyroSAR/tests/test_snap.py:1: in <module>
    from pyroSAR.snap import geocode
pyroSAR/snap/__init__.py:4: in <module>
    snap_config = ExamineSnap()
pyroSAR/snap/auxil.py:278: in __init__
    self.__read_config()
pyroSAR/snap/auxil.py:323: in __read_config
    ConfigHandler.add_section('SNAP')
pyroSAR/_dev_config.py:370: in add_section
    self.parser.write(item)
/usr/lib/python3.4/configparser.py:897: in write
    self._sections[section].items(), d)
/usr/lib/python3.4/configparser.py:901: in _write_section
    fp.write("[{}]\n".format(section_name))
E   TypeError: 'str' does not support the buffer interface
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.79 seconds ============================
```

### CODE REVIEW

**Comment by johntruckenbrodt** [comment]
have you checked [this post](https://stackoverflow.com/questions/7139225/typeerror-str-does-not-support-the-buffer-interface) yet?

**Comment by ibaris** [comment]
I tried it this morning and several other options @johntruckenbrodt. The problem is that python 2.7 does not now the 'encoding' parameter. I get this here
```
Traceback (most recent call last):
  File "C:\Users\ibari\Anaconda2\lib\site-packages\IPython\core\interactiveshell.py", line 2878, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-a718b7ef0a8f>", line 1, in <module>
    pyroSAR.ConfigHandler.add_section('URL')
  File "C:\Users\ibari\Dropbox\GitHub\pyroSAR\pyroSAR\_dev_config.py", line 369, in add_section
    with open(self.__GLOBAL['config'], 'wb', encoding='utf8') as item:
TypeError: 'encoding' is an invalid keyword argument for this function
´´´

**Comment by johntruckenbrodt** [comment]
Okay got it. Two things. First, I think the combination of mode 'wb' and encoding 'utf-8' could be problematic even if using Python 3. As I understand it, all strings written to the file will be encoded to utf-8, but the file is open in binary mode. Second, the difference between binary and str is tricky to incorporate supporting both Python 2.7 and 3 in the package. How about doing a version differentiation here:
```Python
import sys
if sys.version_info >= (3, 0):
    with open(self.__GLOBAL['config'], 'w', encoding='utf8') as item:
        ...
else:
    with open(self.__GLOBAL['config'], 'w') as item:
```
Not sure this works in your case though..

**Comment by ibaris** [comment]
That should work now! It has not occurred to me to do it with a condition based on `sys.version_info`. I have to learn a lot due to the cross-platform coding 😃 Thank you very much for your help!

**Comment by johntruckenbrodt** [comment]
Nice one! Don't worry about AppVeyor failing at the moment. I still have to figure out the reason for it. It used to work but now doesn't.

File: pyroSAR/snap/auxil.py [hunk 1]
```diff
@@ -274,22 +275,169 @@ def __init__(self):
...
+            if 'properties' in ConfigHandler.keys('SNAP'):
+                self.properties = ConfigHandler.get('SNAP', 'properties')
+
+                if not os.path.exists(self.properties):
+                    self.__get_properties()
+
+            else:
+                self.__get_properties()
+
+            if 'auxdatapath' in ConfigHandler.keys('SNAP'):
```
Comment (johntruckenbrodt) [inline pyroSAR/snap/auxil.py]: The auxdatapath handling might be easier by using property and setter methods:
```Python
    @property
    def auxdatapath(self):
        if not hasattr(self, '__auxdatapath'):
            path_default = os.path.join(os.path.expanduser('~'), '.snap', 'auxdata')
            if 'auxdatapath' in ConfigHandler.keys('SNAP'):
                path_conf = ConfigHandler.get('SNAP', 'auxdatapath')
                # if the path in config is a directory, set this path to the private 
                # __auxdatapath attribute; if not, use the auxdata setter to directly 
                # write the default directory to the config
                if os.path.exists(path_conf):
                    self.__auxdatapath = path_conf
                else:
                    self.auxdatapath = path_default
            else:
                self.auxdatapath = path_default
        return self.__auxdatapath
    
    @auxdatapath.setter
    def auxdatapath(self, path):
        self.__auxdatapath = path
        ConfigHandler.set('SNAP', 'auxdatapath', path, True)

```

File: pyroSAR/_dev_config.py [hunk 2]
```diff
@@ -133,3 +137,353 @@ def __dir__(self):
...
+
+    Notes
+    -----
+    The syntax is the same as in ConfigParser. Here, keys are called options.
+
+    """
+
+    # ---- Define Global Variables ----
+
+    __KEYS = {
```
[the remark this answers is not in the document]
Comment (ibaris) [reply pyroSAR/_dev_config.py]: Yes. A list would be more suitable!

File: pyroSAR/snap/auxil.py (hunk 1 above)
Comment (ibaris) [reply pyroSAR/snap/auxil.py]: Thank you very much for your efforts. I will change that.

**Pull request merged**